src/backend/distributed/utils/resource_lock.c

```diff
@@ -119,3 +124,58 @@ UnlockJobResource(uint64 jobId, LOCKMODE lockmode)
 
 	LockRelease(&tag, lockmode, sessionLock);
 }
+
+
+/*
+ * LockShards takes shared locks on the metadata and the data of all shards in
+ * shardIntervalList. This prevents concurrent placement changes and concurrent
+ * DML statements that require an exclusive lock.
+ */
+void
+LockShards(List *shardIntervalList)
+{
+	ListCell *shardIntervalCell = NULL;
+
+	/* lock shards in order of shard id to prevent deadlock */
+	shardIntervalList = SortList(shardIntervalList, CompareShardIntervalsById);
+
+	foreach(shardIntervalCell, shardIntervalList)
+	{
+		ShardInterval *shardInterval = (ShardInterval *) lfirst(shardIntervalCell);
+		int64 shardId = shardInterval->shardId;
+
+		/* prevent concurrent changes to number of placements */
+		LockShardDistributionMetadata(shardId, ShareLock);
+
+		/* prevent concurrent update/delete statements */
+		LockShardResource(shardId, ShareLock);
```

Comment:
Changed the signature of LockShards function so it takes a locktype. It is ShareLock in copy and ExlusiveLock in multi shard modify.